src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -125,102 +123,198 @@ private static List<LoginMethod> getParsedLoginMethods(Main main, AppIdentifier
+        // We don't perform any normalisation on the period in ImportTotpDeviceAPI.java other than checking if it is >= 0
+        if (skew != null && skew.intValue() < 0) {
+            errors.add("skew should be >= 0 for a totp device.");
+            return null;
+        }
+        return skew != null ? skew.intValue() : null;
+    }
+
+    private static String validateAndNormaliseTotpDeviceName(String deviceName) {
+        // We normalise the deviceName as per the ImportTotpDeviceAPI.java
+        return deviceName != null ? deviceName.trim() : null;
+    }
+
+    private static void validateAndNormaliseIsPrimaryField(JsonArray jsonLoginMethods, List<String> errors) {
+        // We are validating that only one loginMethod has isPrimary as true
+        boolean hasPrimaryLoginMethod = false;
+        for (JsonElement jsonLoginMethod : jsonLoginMethods) {
+            JsonObject jsonLoginMethodObj = jsonLoginMethod.getAsJsonObject();
+            if (validateJsonFieldType(jsonLoginMethodObj, "isPrimary", ValueType.BOOLEAN)) {
+                if (jsonLoginMethodObj.get("isPrimary").getAsBoolean()) {
+                    if (hasPrimaryLoginMethod) {
+                        errors.add("No two loginMethods can have isPrimary as true.");
+                    }
+                    hasPrimaryLoginMethod = true;
+                }
+            }
+        }
+    }
+
+    private static String validateAndNormaliseRecipeId(String recipeId, List<String> errors) {
+        if (recipeId == null) {
+            return null;
+        }
+
+        // We don't perform any normalisation on the recipeId after reading it from request header.
+        // We will validate it as is.
+        if (!Arrays.asList("emailpassword", "thirdparty", "passwordless").contains(recipeId)) {
+            errors.add("Invalid recipeId for loginMethod. Pass one of emailpassword, thirdparty or, passwordless!");
+        }
+        return recipeId;
+    }
+
+    private static String validateAndNormaliseTenantId(Main main, AppIdentifier appIdentifier, String tenantId, String recipeId, List<String> errors)
             throws StorageQueryException, TenantOrAppNotFoundException {
         if (tenantId == null || tenantId.equals(TenantIdentifier.DEFAULT_TENANT_ID)) {
-            return;
+            return tenantId;
         }
 
         if (Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
                 .noneMatch(t -> t == EE_FEATURES.MULTI_TENANCY)) {
             errors.add("Multitenancy must be enabled before importing users to a different tenant.");
-            return;
+            return tenantId;
         }
  
+        // We make the tenantId lowercase while parsing from the request in WebserverAPI.java
+        String normalisedTenantId = tenantId.trim().toLowerCase();
         TenantConfig[] allTenantConfigs = Multitenancy.getAllTenantsForApp(appIdentifier, main);
         ArrayList<String> validTenantIds = new ArrayList<>();
```

Comment:
A Set would be a better data structure for this?

Reply:
Changed it.